src/mpack/mpack-writer.h

```diff
@@ -765,6 +765,82 @@ MPACK_INLINE void mpack_finish_type(mpack_writer_t* writer, mpack_type_t type) {
     mpack_writer_track_pop(writer, type);
 }
 
+#if defined(__cplusplus)
+#define MPACK_HAS_GENERIC 1
```

Comment:
I think `MPACK_HAS_GENERIC` should only be defined if we are sure the compiler supports the `_Generic` keyword, and it should be defined in `mpack-platform.h`. I'd recommend moving the macro version above the C++ fallbacks if the preproc is defined. For example:

```
#if MPACK_HAS_GENERIC
// define mpack_write() with _Generic
#elif defined(__cplusplus)
// define mpack_write() C++ overloads
#endif
```

This could allow us to use `_Generic` on C compilers that don't necessarily advertise C99 support (for example MSVC may add support for it in C mode long before incrementing `__STDC_VERSION__`.)